Upgrade iov-bns for weave v0.10 compatibility

ethanfrey opened the pull request:
Closes #645 

- [x] Improve protoc scripts to ignore gogoproto extensions 
- [x] Compile new protobuf codec
- [x] Fix all compile errors in TypeScript code (js object -> protobuf mapping)
- [x] Fix all BnsConnection tests
- [x] Re-generate test vectors (in testdata.spec.ts) from v0.10 golang code
- [x] Fix all codec tests

webmaster128 commented:
Depends on #730

ethanfrey commented on scripts/bnsd/genesis_app_state.json:
```diff
@@ -12,7 +12,7 @@
     {
       "address": "053e4ffbb040d4dc3b201dfe9264c00d918e087b",
-      "name": "faucet",
+      "// name": "faucet",
```
It is unused, but they ignore unused fields...
Want to be clear no one parses this data ;)

ethanfrey commented:
I have issues getting test vectors build and working at all... basic stuff like the keys (which should be unchanged).

Nonetheless, this works as far as the tests tell, with an actually bnsd v0.10.0 behind tendermint v0.27.0.

It needs some polish with those codec tests, but otherwise, basically ready

webmaster128 commented on packages/iov-bns/.gitignore:
```diff
@@ -1,4 +1,9 @@
+go/
 weave/
+clean/
+testvectors/
```
Newline missing. Where did your editorconfig plugin go? However, not blocking

webmaster128 commented on packages/iov-bns/scripts/build_bnsd.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Shell header missing

```
set -o errexit -o nounset -o pipefail
command -v shellcheck > /dev/null && shellcheck "$0"
```

webmaster128 commented on packages/iov-bns/scripts/clean.sh:
```diff
@@ -0,0 +1,30 @@
+  # This just removes all options after the fields, and removes illegal ;; typos
+  cat $file | sed 's/ *\[[^]]*\];/;/g' | sed 's/;;/;/' > $outfile
+done
+cd -
```
Pro tip for path management in shell script: `cd` in a subshell:

```shell
echo "Original: $(pwd)"
(
  cd "some/sub/folder"
  # execute stuff
)
echo "unchanged base: $(pwd)"

webmaster128 commented on packages/iov-bns/scripts/jsonbytes:
```diff
@@ -0,0 +1,7 @@
+# Usage: jsonbytes <filename> <jq selector to field>
+# Extracts the bytes (base64 encoded) and converts to hex
+
+cat "$1" | jq -r "$2" | base64 -d - | cat - | od -x | cut -c9- | tr -d ' \n'
```
1. Unnecessary cat: `jq -r "$2" "$1"`
1. Does `base64 -d` work on Linux? On mac it is `-D, --decode   decodes input`
2. `cat -` is a no-op, right?
3. I think the byte order is not correct in `| od -x | cut -c9-`. Confirm `echo abc | xxd -p` vs. `echo abc | od -x | cut -c9-`. You can use `xxd -p -c 999999999` to avoid line breaks.

ethanfrey replied:
wow!
`I think the byte order is not correct in | od -x | cut -c9-. Confirm echo abc | xxd -p vs. echo abc | od -x | cut -c9-. You can use xxd -p -c 999999999 to avoid line breaks.`

I will fix scripts.

Thank you so much!

ethanfrey replied:
2. `base64 -d` is linux, `base64 -D` for osx :disappointed: 
Can you see if `base64 --decode` also works on osx, as a generic alternative?

webmaster128 replied:
> Can you see if base64 --decode also works on osx, as a generic alternative?

It is documented, so yeah, it will do the job

ethanfrey replied:
`xxd -p -c 9999` tip is gold :+1:

webmaster128 commented on packages/iov-bns/scripts/jsonbytes:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
shell header missing

webmaster128 commented on packages/iov-bns/scripts/tohex:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Shell header missing

webmaster128 commented on packages/iov-bns/scripts/tohex:
```diff
@@ -0,0 +1,7 @@
+# take input from a file if given, otherwise stdin
+IN=${1:--}
+
+cat $IN | od -x | cut -c9- | tr -d ' \n'
```
Same as above: incorrect hex

webmaster128 commented on packages/iov-bns/.gitignore:
```diff
@@ -1,4 +1,9 @@
+
+# build dirs for creating protobuf definitions
+go/
 weave/
```
`packages/iov-bns/weave/` is not used anymore, right? Then this line should go

webmaster128 commented on packages/iov-bns/src/bnscodec.spec.ts:
```diff
@@ -15,7 +15,7 @@ import {
 } from "./testdata.spec";
 
-describe("bnscodec", () => {
+xdescribe("bnscodec", () => {
```
This should not be checked in

webmaster128 commented on packages/iov-bns/src/bnsconnection.spec.ts:
```diff
@@ -882,6 +891,7 @@ describe("BnsConnection", () => {
       const signed = await profile.signTransaction(mainWalletId, faucet, sendTx, bnsCodec, nonce);
       const response = await connection.postTx(bnsCodec.bytesToPost(signed));
       const blockInfo = await response.blockInfo.waitFor(info => !isBlockInfoPending(info));
+      expect(blockInfo.state).toBe(TransactionState.Succeeded);
```
Pease use toEqual for all equality checks for consistency and readability

webmaster128 commented on packages/iov-bns/src/bnsconnection.ts:
```diff
@@ -668,7 +668,7 @@ export class BnsConnection implements BcpAtomicSwapConnection {
       results = (await this.query("/nft/usernames/owner", rawAddress)).results;
     } else if (isBnsUsernamesByChainAndAddressQuery(query)) {
-      const pairSerialized = `${query.address}*${query.chain}`;
+      const pairSerialized = `${query.address};${query.chain}`;
```
😍😍😍😍😍

webmaster128 commented on packages/iov-bns/src/decode.spec.ts:
```diff
@@ -216,7 +216,7 @@ describe("Decode", () => {
   });
 
-  describe("transactions", () => {
+  xdescribe("transactions", () => {
```
should not be checked in

webmaster128 commented on packages/iov-bns/src/decode.spec.ts:
```diff
@@ -236,7 +236,7 @@ describe("Decode", () => {
   });
 
-  describe("parseMsg", () => {
+  xdescribe("parseMsg", () => {
```
should not be checked in

webmaster128 commented on packages/iov-bns/src/encode.spec.ts:
```diff
@@ -138,7 +138,7 @@ describe("Encode", () => {
   });
 
-  describe("buildMsg", () => {
+  xdescribe("buildMsg", () => {
```
should not be checked in

webmaster128 commented on packages/iov-bns/src/encode.spec.ts:
```diff
@@ -235,14 +235,14 @@ describe("Encode", () => {
 });
 
-describe("Encode transactions", () => {
+xdescribe("Encode transactions", () => {
```
should not be checked in

webmaster128 commented on scripts/bnsd/start.sh:
```diff
@@ -3,9 +3,9 @@ set -o errexit -o nounset -o pipefail
+export BNSD_TM_VERSION=v0.27.4
 # Choose from https://hub.docker.com/r/iov1/bnsd/tags/
-export BNSD_VERSION=v0.9.2
+export BNSD_VERSION=v0.10.0
```
Why not v0.10.2?

ethanfrey commented:
I think I addressed all PR comments, and now have the proper byte order in test vectors.
All tests re-enabled and passing (also the minHeight/maxHeight one)

ethanfrey commented:
Okay, this passes now, except for failures in iov-faucets.
I remember this chicken-and-egg problems before, as the faucets needs v0.12 release in order to work, but the tests can pass to release until the faucet works...

I don't remember what you did last time, but can you tell me the proper bootstrap technique (just disable tests temporarily?). Then we should be green

```
lerna ERR! yarn run test exited 1 in '@iov/faucets'
lerna ERR! yarn run test stdout:
$ yarn build-or-skip && yarn test-node
$ [ -n "$SKIP_BUILD" ] || yarn build
$ node jasmine-testrunner.js
Executing 5 defined specs...
Test Suites & Specs:
1. IovFaucet
-    can be constructed
   ✔ can be constructed (3ms)
-    can be used to credit a wallet
   ✖ can be used to credit a wallet (1 failure) (17ms)
-    can be used to credit a wallet with a different token
   ✖ can be used to credit a wallet with a different token (1 failure) (3ms)
-    throws for invalid ticker
   ✖ throws for invalid ticker (1 failure) (8ms)
-    throws for invalid address
   ✖ throws for invalid address (2 failures) (3ms)
```

webmaster128 commented on packages/iov-bns/src/testdata.spec.ts:
```diff
@@ -25,80 +25,104 @@ const { fromHex } = Encoding;
+// bech32 -e -h tiov $ADDR
+export const address = "tiov1mkp7tg2qvyp6m3uttw44ju9dhjzk6mdgzcz6yf" as Address;
 
+// TODO??? where is this from????
```
"from new testgen" ;) https://github.com/iov-one/iov-core/commit/c376649881fbba0f8cd43293dd485e83f1be67e8

webmaster128 commented on packages/iov-bns/scripts/tohex:
```diff
@@ -0,0 +1,7 @@
+
+# take input from a file if given, otherwise stdin
+IN=${1:--}
+xxd -p -c 999999 < "$IN"
```
`xxd -p -c 999999 < "-"` will not do what you want. Do we need both features?

ethanfrey replied:
Fair enough, will simplify

webmaster128 commented on packages/iov-bns/src/bnsconnection.spec.ts:
```diff
@@ -937,8 +947,9 @@ describe("BnsConnection", () => {
-    xit("can search for transactions by minHeight/maxHeight", async () => {
+    // Fixed since tendermint v0.26.4
+    // see issue https://github.com/tendermint/tendermint/issues/2759
+    it("can search for transactions by minHeight/maxHeight", async () => {
```
Oh, very nice!

webmaster128 commented on packages/iov-bns/PROTOBUF.md:
```diff
@@ -44,4 +44,49 @@ To pull this all together, you can just use the helper function `protoc`... This
+
+There is often a bit of adjusting to adapt the json to the internal object format, but it should be straight-forward.
+
+Note, you will also have to compile and install this script as a bech32 helper (until we add native support for this in iov-core):
```
I think IOV-Core should never be used to generate test data to test IOV-Core. So using an external bech32 tool is a solid, long-term stategy.

ethanfrey commented:
Thanks.
Appveyor died due to rise timeout issue, don't know how to restart the job

webmaster128 commented:
>  Appveyor died due to rise timeout issue, don't know how to restart the job

Only I can do that with my personal account. If you really need to do that one time, you can close and repoen the PR. Will take care of it now
